internal/hcsoci/cmd.go

```diff
@@ -0,0 +1,318 @@
+func (c *Cmd) Start() error {
+	c.allDoneCh = make(chan struct{})
+	var x interface{}
+	if c.OS == "windows" || c.InUVM {
+		wpp := &hcsschema.ProcessParameters{
+			CommandLine:      c.Spec.CommandLine,
+			User:             c.Spec.User.Username,
+			WorkingDirectory: c.Spec.Cwd,
+			EmulateConsole:   c.Spec.Terminal,
+			CreateStdInPipe:  c.Stdin != nil,
+			CreateStdOutPipe: c.Stdout != nil,
+			CreateStdErrPipe: c.Stderr != nil,
+		}
+
+		if c.Spec.CommandLine == "" {
+			if c.OS == "windows" {
+				wpp.CommandLine = escapeArgs(c.Spec.Args)
+			} else {
+				wpp.CommandArgs = c.Spec.Args
+			}
+		}
+
+		environment := make(map[string]string)
+		for _, v := range c.Spec.Env {
+			s := strings.SplitN(v, "=", 2)
+			if len(s) == 2 && len(s[1]) > 0 {
+				environment[s[0]] = s[1]
+			}
+		}
+		wpp.Environment = environment
+
+		if c.Spec.ConsoleSize != nil {
+			wpp.ConsoleSize = []int32{
+				int32(c.Spec.ConsoleSize.Height),
+				int32(c.Spec.ConsoleSize.Width),
+			}
+		}
+		x = wpp
+
+	} else {
+		lpp := &lcowProcessParameters{
+			ProcessParameters: hcsschema.ProcessParameters{
+				CreateStdInPipe:  c.Stdin != nil,
+				CreateStdOutPipe: c.Stdout != nil,
+				CreateStdErrPipe: c.Stderr != nil,
+			},
+			OCIProcess: c.Spec,
+		}
+		x = lpp
+	}
+	if c.Context != nil && c.Context.Err() != nil {
+		return c.Context.Err()
+	}
+	p, err := c.Host.CreateProcess(x)
+	if err != nil {
+		return err
+	}
+	c.Process = p
+	if c.Log != nil {
+		c.Log = c.Log.WithField("pid", p.Pid())
+	}
+
+	// Start relaying process IO.
+	stdin, stdout, stderr := p.Stdio()
+	if c.Stdin != nil {
+		// Do not make stdin part of the error group because there is no way for
+		// us or the caller to reliably unblock the c.Stdin read when the
+		// process exits.
+		go func() {
+			_, err := copyAndLog(stdin, c.Stdin, c.Log, "stdin")
+			// Report the stdin copy error. If the process has exited, then the
+			// caller may never see it, but if the error was due to a failure in
+			// stdin read, then it is likely the process is still running.
+			if err != nil {
+				c.stdinErr.Store(err)
+			}
+			// Notify the process that there is no more input.
+			p.CloseStdin()
+		}()
+	}
+
+	if c.Stdout != nil {
+		c.iogrp.Go(func() error {
+			_, err := copyAndLog(c.Stdout, stdout, c.Log, "stdout")
+			return err
+		})
+	}
+
+	if c.Stderr != nil {
+		c.iogrp.Go(func() error {
+			_, err := copyAndLog(c.Stderr, stderr, c.Log, "stderr")
+			return err
+		})
+	}
+
+	if c.Context != nil {
+		go func() {
+			select {
+			case <-c.Context.Done():
+				c.Process.Kill()
+			case <-c.allDoneCh:
+			}
+		}()
+	}
+	return nil
+}
+
+// Wait waits for a command and its IO to complete and closes the underlying
+// process. It can only be called once. It returns an ExitError if the command
+// runs and returns a non-zero exit code.
+func (c *Cmd) Wait() error {
+	err := c.Process.Wait()
+	if err != nil && c.Log != nil {
+		c.Log.WithError(err).Warn("process wait failed")
```

Comment:
I don't remember. Does this place `logrus.ErrorKey` in the `Data` map or does this make a copy? Wouldn't want all future log messages to have this error

Reply:
WithError returns a new entry.

Reply:
(with new fields)